modules/core/src/main/java/org/locationtech/jts/index/kdtree/KdTree.java

```diff
@@ -179,6 +180,180 @@ public KdNode insert(Coordinate p, Object data) {
     
     return insertExact(p, data);
   }
+  
+  /**
+   * Finds the nearest node in the tree to the given query point.
+   * 
+   * @param query the query point
+   * @return the nearest node, or null if the tree is empty
+   */
+  public KdNode nearestNeighbor(final Coordinate query) {
+    if (root == null) {
+      return null;
+    }
+
+    KdNode bestNode = null;
+    double bestDistance = Double.POSITIVE_INFINITY;
+    Deque<NNStackFrame> stack = new ArrayDeque<>();
+    KdNode currentNode = root;
+    boolean isXLevel = true;
+
+    while (currentNode != null || !stack.isEmpty()) {
+      if (currentNode != null) {
+        double currentDist = query.distanceSq(currentNode.getCoordinate());
+        if (currentDist < bestDistance) {
+          bestNode = currentNode;
+          bestDistance = currentDist;
+          if (bestDistance == 0) {
+              return bestNode; // Early termination
+          }
+        }
+
+        boolean currentIsXLevel = isXLevel;
+        double splitValue = currentNode.splitValue(currentIsXLevel);
+        KdNode nextNode;
+        KdNode otherNode;
+
+        if (currentIsXLevel) {
+          if (query.x < splitValue) {
+            nextNode = currentNode.getLeft();
+            otherNode = currentNode.getRight();
+          } else {
+            nextNode = currentNode.getRight();
+            otherNode = currentNode.getLeft();
+          }
+        } else {
+          if (query.y < splitValue) {
+            nextNode = currentNode.getLeft();
+            otherNode = currentNode.getRight();
+          } else {
+            nextNode = currentNode.getRight();
+            otherNode = currentNode.getLeft();
+          }
+        }
+
+        stack.push(new NNStackFrame(otherNode, currentIsXLevel, splitValue));
+        currentNode = nextNode;
+        isXLevel = !currentIsXLevel;
+      } else {
+        NNStackFrame frame = stack.pop();
+        KdNode otherNode = frame.node;
+        boolean parentSplitAxis = frame.parentSplitAxis;
+        double parentSplitValue = frame.parentSplitValue;
+
+        double diff = parentSplitAxis 
+        	    ? query.x - parentSplitValue
+        	    : query.y - parentSplitValue;
+        double distanceToSplitSq = diff * diff;
+
+        if (distanceToSplitSq < bestDistance) {
+          currentNode = otherNode;
+          isXLevel = !parentSplitAxis;
+        } else {
+          currentNode = null;
+        }
+      }
+    }
+
+    return bestNode;
+  }
+  
+  /**
+   * Finds the nearest N nodes in the tree to the given query point.
+   * 
+   * @param query the query point
+   * @param n the number of nearest nodes to find
+   * @return a list of the nearest nodes, sorted by distance (closest first), or an empty list if the tree is empty.
+   */
+  public List<KdNode> nearestNeighbors(final Coordinate query, final int n) {
```

Comment:
These method names are different to the ones on `STRtree` - e.g. it has `nearestNeighbourK`.  But I prefer this naming style.  Perhaps `STRtree` should add methods with these names and deprecate the old ones.

Reply:
Are you suggesting to do that in this PR?

Reply:
No, they can be added in a future PR.

Reply:
Ok, ready for merge then I think.